cpp/src/structs/utilities.cpp

```diff
@@ -168,15 +167,132 @@ flatten_nested_columns(table_view const& input,
-  if (not has_struct)
-    return std::make_tuple(input, column_order, null_precedence, std::move(validity_as_column));
+  if (not has_struct) {
+    return std::make_tuple(
+      input, column_order, null_precedence, std::vector<std::unique_ptr<column>>{});
+  }
 
   return flattened_table{input, column_order, null_precedence, nullability}();
 }
 
+using vector_of_columns = std::vector<std::unique_ptr<cudf::column>>;
+using column_index_t    = typename vector_of_columns::size_type;
+
+namespace {
+// Forward declaration, to enable recursion via `unflattener`.
+std::unique_ptr<cudf::column> unflatten_struct(vector_of_columns& flattened,
+                                               column_index_t& current_index,
+                                               cudf::column_view const& blueprint);
+
+/**
+ * @brief Helper functor to reconstruct STRUCT columns from its flattened member columns.
+ *
+ */
+class unflattener {
+ public:
+  unflattener(vector_of_columns& flattened_, column_index_t& current_index_)
+    : flattened{flattened_}, current_index{current_index_}
+  {
+  }
+
+  auto operator()(column_view const& blueprint)
+  {
+    return blueprint.type().id() == type_id::STRUCT
+             ? unflatten_struct(flattened, current_index, blueprint)
+             : std::move(flattened[current_index++]);
+  }
+
+ private:
+  vector_of_columns& flattened;
+  column_index_t& current_index;
+
+};  // class unflattener;
+
+std::unique_ptr<cudf::column> unflatten_struct(vector_of_columns& flattened,
+                                               column_index_t& current_index,
+                                               cudf::column_view const& blueprint)
+{
+  // "Consume" columns from `flattened`, starting at `current_index`,
+  // based on the provided `blueprint` struct col. Recurse for struct children.
+  CUDF_EXPECTS(blueprint.type().id() == type_id::STRUCT,
+               "Expected blueprint column to be a STRUCT column.");
+
+  CUDF_EXPECTS(current_index < flattened.size(), "STRUCT column can't have 0 children.");
+
+  auto const num_rows = flattened[current_index]->size();
+
+  // cudf::flatten_nested_columns() executes depth first, and serializes the struct null vector
+  // before the child/member columns.
+  // E.g. STRUCT_1< STRUCT_2< A, B >, C > is flattened to:
+  //      1. Null Vector for STRUCT_1
+  //      2. Null Vector for STRUCT_2
+  //      3. Member STRUCT_2::A
+  //      4. Member STRUCT_2::B
+  //      5. Member STRUCT_1::C
+  //
+  // Extract null-vector *before* child columns are constructed.
+  auto struct_null_column_contents = flattened[current_index++]->release();
+
+  auto struct_members = vector_of_columns{};
+  struct_members.reserve(blueprint.num_children());
+
+  std::transform(blueprint.child_begin(),
+                 blueprint.child_end(),
+                 std::back_inserter(struct_members),
+                 unflattener{flattened, current_index});
```

Comment:
I just have a cleaner initializing of `std::vector`, so you can have a look: https://github.com/rapidsai/cudf/pull/9012/files#diff-4e8568c7e64b447e404839e2f3b5c97aac92031e1e05e2d96edcfec0e0c61c65R87-R91

Reply:
Oh, yes. Thanks! This is a much better way of doing it. I should just use the iterator-based constructor.